dev-docs/RFCs/v1.0/generic-edit-mode.md

```diff
@@ -0,0 +1,200 @@
+
+  // Callback used to notify applications of an edit action
+  onEdit: (editAction: EditAction<TData>) => void,
+
+  // Callback used to update guides
+  onUpdateGuides: (guides: ?TGuides) => void,
+
+  // Callback used to update cursor
+  onUpdateCursor: (cursor: string) => void
+};
+
+export class EditMode<TData, TGuides> {
+  state: ModeState<TData, TGuides>;
+
+  getState(): ModeState<TData, TGuides> {
+    return this.state;
+  }
+
+  updateState(state: ModeState<TData, TGuides>) {
+    const changedEvents: (() => void)[] = [];
+    if (this.state && this.state.data !== state.data) {
+      changedEvents.push(this.onDataChanged);
+    }
+    if (this.state && this.state.modeConfig !== state.modeConfig) {
+      changedEvents.push(this.onModeConfigChanged);
+    }
+    if (this.state && this.state.selectedIndexes !== state.selectedIndexes) {
+      changedEvents.push(this.onSelectedIndexesChanged);
+    }
+    if (this.state && this.state.guides !== state.guides) {
+      changedEvents.push(this.onGuidesChanged);
+    }
+    this.state = state;
+
+    changedEvents.forEach(fn => fn.bind(this)());
+  }
+
+  // Overridable user interaction handlers
+  handleClick(event: ClickEvent): void {}
+  handlePointerMove(event: PointerMoveEvent): void {}
+  handleStartDragging(event: StartDraggingEvent): void {}
+  handleStopDragging(event: StopDraggingEvent): void {}
+
+  // Convenience functions to handle state changes
+  onDataChanged(): void {}
+  onModeConfigChanged(): void {}
+  onSelectedIndexesChanged(): void {}
+  onGuidesChanged(): void {}
+
+  // Convenience functions to access state
+  getModeConfig(): any {
+    return this.state.modeConfig;
+  }
+  getSelectedIndexes(): number[] {
+    return this.state.selectedIndexes;
+  }
+  getGuides(): ?TGuides {
+    return this.state && this.state.guides;
+  }
+  onEdit(editAction: EditAction<TData>): void {
+    this.state.onEdit(editAction);
+  }
+  onUpdateGuides(guides: ?TGuides): void {
+    this.state.onUpdateGuides(guides);
+  }
+  onUpdateCursor(cursor: string): void {
+    this.state.onUpdateCursor(cursor);
+  }
+}
+```
+
+A user of this class will need to call `updateState` anytime the data within `ModeState` change. This is similar to how React calls `render` and how deck.gl calls each layer's `updateState` function.
+
+An implementation of a mode is intended to override the `handle...` functions in order to handle user input. The mode can then call the callbacks provided in `ModeState` (e.g. `onEdit`).
+
+## Integration to react-map-gl-draw
+
+TODO
+
+## Integration with nebula
+
+### Module layout
+
+We will need a `@nebula.gl/core` module separate from `nebula.gl` module. The reason is because this new `@nebula/core` should have no deck.gl dependency.
+
+* `nebula.gl`
+  * depends on `@nebula.gl/core`, `@nebula.gl/layers`, and all the other `@nebula/...` modules.
+  * doesn't have anything in it, just basically imports from the others and re-exports them
+* `@nebula.gl/core`
```

Comment:
I think this can go into `@nebula.gl/geojson-modes` later we can extract it.
Don't want too many modules.